### PR TITLE
feat(redis): add asyncio support of redis >= 4.2

### PR DESCRIPTION
**Pull request by sileht**
# Integration

This PR adds support for asyncio part of redis >= 4.2

## Links

https://github.com/redis/redis-py#python-notice
https://redis-py.readthedocs.io/en/stable/connections.html#async-client
https://github.com/redis/redis-py/tree/master/redis/asyncio

- Integration docs: ???
- Corp docs PR: ???

## Checklist

- [ ] Usage and configuration documentation added in `__init__.py`, `docs/index.rst` and `docs/integrations.rst`.
- [ ] [Corp docs](https://github.com/Datadog/documentation) PR to add new integration to documentation.
- [x] Span metadata
  - [x] Service (use [`int_service`](https://github.com/DataDog/dd-trace-py/blob/90d1d5981c72ea312c21ac04e5be47521d0f0f2e/ddtrace/contrib/trace_utils.py#L55) or [`ext_service`](https://github.com/DataDog/dd-trace-py/blob/90d1d5981c72ea312c21ac04e5be47521d0f0f2e/ddtrace/contrib/trace_utils.py#L87)).
  - [x] Span type should be one of [these](https://github.com/DataDog/dd-trace-py/blob/90d1d5981c72ea312c21ac04e5be47521d0f0f2e/ddtrace/ext/__init__.py#L7).
  - [x] Resource
  - [x] Measured tag
  - [x] Sample analytics
- [x] Global configuration
  - [x] `ddtrace.config` entry is specified.
  - [x] Environment variables are provided for config options.
- [x] Instance configuration
  - [x] Pin overriding.
  - [x] Service name override (if applicable).
- [ ] Async
  - [ ] Span parenting behaves as expected.
  - [ ] Context propagation across async contexts.
- [ ] HTTP
  - [ ] Distributed tracing propagation is implemented.
  - [ ] Use [`trace_utils.set_http_meta`](https://github.com/DataDog/dd-trace-py/blob/90d1d5981c72ea312c21ac04e5be47521d0f0f2e/ddtrace/contrib/trace_utils.py#L143-L152) to set http tags
- [x] Tests
  - [x] Use `pytest` fixtures found in `tests/conftest.py` or the test helpers on `TracerTestCase` if
    writing `unittest` style test cases.
  - [x] Tests are provided for all the above.
  - [x] Tests are added to CI (`.circleci/config.yml`).
  - [x] Functionality is maintained from original library.
  - [x] Patch test cases are added (see `test_django_patch.py` for an example).
  - [x] All Python versions that the library supports are tested.
  - [x] All significant library versions (including the latest) are tested. This typically includes every minor release going back a few years.

### CODE REVIEW

File: tests/contrib/redis/test_redis_asyncio.py [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+    assert response_list1[0] is True  # response from redis.set is OK if successfully pushed
+    assert response_list2[0] is True
+    assert (
+        response_list1[1].decode() == "boo"
+    )  # response from hset is 'Integer reply: The number of fields that were added.'
+    assert response_list2[1].decode() == "bar"
+
+
+@pytest.mark.asyncio
+async def test_parenting(redis_client):
```
Comment (jd) [inline tests/contrib/redis/test_redis_asyncio.py:211]: do you want to use the snapshot here too? Wondering since there's no assertion otherwise.

Comment (sileht) [reply tests/contrib/redis/test_redis_asyncio.py:213]: Fixed.

Comment (jd) [reply tests/contrib/redis/test_redis_asyncio.py:213]: I'm not seeing any change?

Comment (sileht) [reply tests/contrib/redis/test_redis_asyncio.py:215]: Fixed.

**Comment by sileht** [comment]
I need some insight to ensure the test_redis_asyncio.py file is skipped when the redis integration is tested against python 2.7?

**Comment by sileht** [comment]
[brettlangdon](https://github.com/brettlangdon)

> I need some insight to ensure the test_redis_asyncio.py file is skipped when the redis integration is tested against python 2.7?

Found it!

**Comment by sileht** [comment]
@brettlangdon @Kyle-Verhoog Can I get a review?

Remaining CIs failure seems unrelated to my change.

**Review by jd — DISMISSED** [review]

**Comment by sileht** [comment]
@brettlangdon @Kyle-Verhoog Can I get a review?

File: ddtrace/contrib/redis/asyncio_patch.py [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+from ...internal.utils.formats import stringify_cache_args
+from ...pin import Pin
+from .util import _trace_redis_cmd
+from .util import _trace_redis_execute_pipeline
+
+
+#
+# tracing async functions
+#
+async def traced_async_execute_command(func, instance, args, kwargs):
```
Comment (Kyle-Verhoog) [inline ddtrace/contrib/redis/asyncio_patch.py:12]: https://github.com/DataDog/dd-trace-py/blob/0d62b96977545584c3204cfad2cbebc9c959ec1d/ddtrace/contrib/trace_utils_async.py#L13-L20 should be used here to deal with the pin boilerplate

File: ddtrace/contrib/redis/asyncio_patch.py [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+    pin = Pin.get_from(instance)
+    if not pin or not pin.enabled():
+        return await func(*args, **kwargs)
+
+    with _trace_redis_cmd(pin, config.redis, instance, args):
+        return await func(*args, **kwargs)
+
+
+async def traced_async_execute_pipeline(func, instance, args, kwargs):
```
Comment (Kyle-Verhoog) [inline ddtrace/contrib/redis/asyncio_patch.py:21]: same here

File: tests/contrib/redis/test_redis_asyncio.py [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+    val_list = await redis_client.mget(*range(length))
+    assert len(val_list) == length
+    for val in val_list:
+        assert val is None
+
+
+@pytest.mark.asyncio
+@pytest.mark.snapshot
+async def test_override_service_name(redis_client):
+    with override_config("aioredis", dict(service_name="myaioredis")):
```
Comment (Kyle-Verhoog) [inline tests/contrib/redis/test_redis_asyncio.py:113]: 🤔 this test should be failing/not applicable

Comment (Kyle-Verhoog) [inline tests/contrib/redis/test_redis_asyncio.py:113]:
```suggestion
    with override_config("redis", dict(service_name="myredis")):
```

**Review by Kyle-Verhoog — COMMENTED** [review]
Just a few smaller things otherwise lgtm

File: tests/contrib/redis/test_redis_asyncio.py (hunk 4 above)
Comment (sileht) [reply tests/contrib/redis/test_redis_asyncio.py]: fixed

File: ddtrace/contrib/redis/asyncio_patch.py (hunk 2 above)
Comment (sileht) [reply ddtrace/contrib/redis/asyncio_patch.py:12]: @Kyle-Verhoog WDYT?

**Review by jd — APPROVED** [review]

**Review by brettlangdon — APPROVED** [review]

**Pull request merged**